File: packages/sandbox/daemon/setup/dep-metrics.test.ts
import { describe, expect, it } from "bun:test";
import { buildDepLines, isPackageManifest } from "./dep-metrics";

describe("isPackageManifest", () => {
  it("accepts real package roots (npm/yarn/bun + nested + pnpm)", () => {
    expect(isPackageManifest("react/package.json")).toBe(true);
    expect(isPackageManifest("@scope/pkg/package.json")).toBe(true);
    expect(isPackageManifest("foo/node_modules/bar/package.json")).toBe(true);
    expect(
      isPackageManifest(".pnpm/bar@1.0.0/node_modules/bar/package.json"),
    ).toBe(true);
    expect(
      isPackageManifest(
        ".pnpm/@scope+bar@1.0.0/node_modules/@scope/bar/package.json",
      ),
    ).toBe(true);
  });

  it("rejects fixture/sample package.json shipped inside a package", () => {
    expect(isPackageManifest("foo/test/fixtures/package.json")).toBe(false);
    expect(isPackageManifest(".cache/foo/package.json")).toBe(false);
    expect(isPackageManifest("package.json")).toBe(false);
  });
});

describe("buildDepLines", () => {
  const input = {
    bootId: "boot-1",
    packageManager: "bun" as const,
    repoName: "org/repo",
    branch: "main",
  };
  const dep = (i: number) => ({
    name: `@scope/package-name-${i}`,
    version: `10.${i}.${i}`,
  });

  // Every emitted line must round-trip and stay under the pipeline's 16KB
  // byte cap; the full dep set must survive across chunks exactly once.
  const assertIntact = (
    lines: string[],
    deps: { name: string; version: string }[],
  ) => {
    const seen: string[] = [];
    for (const line of lines) {
      expect(Buffer.byteLength(line)).toBeLessThan(16_000);
      const o = JSON.parse(line);
      expect(o.msg).toBe("sandbox.deps");
      expect(o.chunks).toBe(lines.length);
      expect(o.dependencyCount).toBe(deps.length);
      expect(typeof o.deps).toBe("string");
      seen.push(...JSON.parse(o.deps));
    }
    expect(seen.length).toBe(deps.length);
    expect(new Set(seen).size).toBe(deps.length);
    return seen;
  };

  it("chunks a big install into parseable lines under the 16KB pipeline cap", () => {
    const deps = Array.from({ length: 391 }, (_, i) => dep(i));
    const lines = buildDepLines(deps, input);
    expect(lines.length).toBeGreaterThan(1);
    const seen = assertIntact(lines, deps);
    expect(seen[0]).toBe("@scope/package-name-0@10.0.0");
  });

  // The bug this fix closes: count-based chunking (200/line) would emit a
  // ~40KB line here and get it truncated. Byte-based chunking must not.
  it("keeps lines under cap even with pathologically long package names", () => {
    const longName = `@${"o".repeat(100)}/${"p".repeat(100)}`; // ~203 chars
    const deps = Array.from({ length: 200 }, (_, i) => ({
      name: longName,
      version: `1.0.${i}`,
    }));
    const lines = buildDepLines(deps, input);
    expect(lines.length).toBeGreaterThan(2);
    assertIntact(lines, deps);
  });

  it("emits a single countable line for a zero-dep install", () => {
    const lines = buildDepLines([], input);
    expect(lines.length).toBe(1);
    const o = JSON.parse(lines[0]);
    expect(o.dependencyCount).toBe(0);
    expect(JSON.parse(o.deps)).toEqual([]);
  });
});

File: packages/sandbox/daemon-go/internal/setup/depmetrics_test.go
package setup

import (
	"encoding/json"
	"strings"
	"testing"
)

func TestBuildDepsRestoreLineShape(t *testing.T) {
	line := BuildDepsRestoreLine(RestoreMiss, "https://u:p@github.com/o/r.git", 1234, "boot-1")
	// Field order must match the TS emitter so the stored lines look identical.
	if !strings.HasPrefix(line, `{"msg":"sandbox.deps.restore","source":"miss","repo_hash":"`) {
		t.Fatalf("unexpected prefix: %s", line)
	}
	var got map[string]any
	if err := json.Unmarshal([]byte(line), &got); err != nil {
		t.Fatal(err)
	}
	if got["duration_ms"] != float64(1234) || got["bootId"] != "boot-1" {
		t.Fatalf("got %v", got)
	}
	// Credentials are stripped before hashing, never emitted.
	if strings.Contains(line, "u:p") || strings.Contains(line, "github.com") {
		t.Fatalf("line leaked the clone URL: %s", line)
	}
	hash, _ := got["repo_hash"].(string)
	if len(hash) != 16 {
		t.Fatalf("repo_hash = %q, want 16 hex chars", hash)
	}
	// The same repo with and without credentials must hash identically.
	bare := BuildDepsRestoreLine(RestoreMiss, "https://github.com/o/r.git", 1, "b")
	var bareGot map[string]any
	json.Unmarshal([]byte(bare), &bareGot)
	if bareGot["repo_hash"] != hash {
		t.Fatalf("credential stripping changed the hash: %v vs %v", bareGot["repo_hash"], hash)
	}
}

func TestBuildDepsRestoreLineUnknownRepo(t *testing.T) {
	line := BuildDepsRestoreLine(RestoreNoInstall, "", 0, "b")
	if !strings.Contains(line, `"repo_hash":"unknown"`) {
		t.Fatalf("got %s", line)
	}
}

func TestIsPackageManifest(t *testing.T) {
	cases := map[string]bool{
		"pkg/package.json":                            true,
		"@scope/pkg/package.json":                     true,
		"pkg/node_modules/nested/package.json":        true,
		"@scope/pkg/node_modules/dep/package.json":    true,
		".pnpm/foo@1.0.0/node_modules/x/package.json": true,
		"pkg/fixtures/sample/package.json":            false,
		"pkg/dist/deep/nested/package.json":           false,
		"package.json":                                false,
	}
	for rel, want := range cases {
		if got := IsPackageManifest(rel); got != want {
			t.Errorf("IsPackageManifest(%q) = %v, want %v", rel, got, want)
		}
	}
}

// Both simpler shapes fail in the log pipeline: one line with the whole array
// is truncated at 16KB, one line per dep gets rate-sampled away.
func TestBuildDepLinesStayUnderByteCap(t *testing.T) {
	flat := make([]string, 400)
	for i := range flat {
		flat[i] = "@scope/some-reasonably-long-package-name-" + string(rune('a'+i%26)) + "@1.2.3"
	}
	lines := BuildDepLines(flat, DepMetricsInput{
		PackageManager: "npm", BootId: "boot-1",
		RepoName: strings.Repeat("r", 200), Branch: strings.Repeat("b", 200),
	})
	if len(lines) < 2 {
		t.Fatalf("expected chunking, got %d line(s)", len(lines))
	}
	seen := 0
	for i, line := range lines {
		if len(line) > maxLineBytes {
			t.Errorf("line %d is %d bytes, over the %d cap", i, len(line), maxLineBytes)
		}
		var parsed struct {
			Chunk, Chunks, DependencyCount int
			Deps                           string
			RepoName, Branch               string
		}
		if err := json.Unmarshal([]byte(line), &parsed); err != nil {
			t.Fatal(err)
		}
		if parsed.Chunks != len(lines) || parsed.Chunk != i+1 {
			t.Errorf("line %d: chunk %d/%d", i, parsed.Chunk, parsed.Chunks)
		}
		if parsed.DependencyCount != len(flat) {
			t.Errorf("dependencyCount = %d, want %d", parsed.DependencyCount, len(flat))
		}
		// Long meta must be clipped, or a single-dep line blows the cap.
		if len(parsed.RepoName) > maxMetaBytes || len(parsed.Branch) > maxMetaBytes {
			t.Errorf("meta not clipped: repo=%d branch=%d", len(parsed.RepoName), len(parsed.Branch))
		}
		// `deps` is a pre-encoded JSON string, not a real array.
		var group []string
		if err := json.Unmarshal([]byte(parsed.Deps), &group); err != nil {
			t.Fatalf("deps is not a JSON-encoded array: %v", err)
		}
		seen += len(group)
	}
	if seen != len(flat) {
		t.Fatalf("chunking lost deps: %d of %d", seen, len(flat))
	}
}

// A zero-dep install still has to emit one countable line, or the denominator
// is unknowable.
func TestBuildDepLinesEmitsForZeroDeps(t *testing.T) {
	lines := BuildDepLines(nil, DepMetricsInput{PackageManager: "npm", BootId: "b"})
	if len(lines) != 1 {
		t.Fatalf("got %d lines, want 1", len(lines))
	}
	if !strings.Contains(lines[0], `"dependencyCount":0`) || !strings.Contains(lines[0], `"deps":"[]"`) {
		t.Fatalf("got %s", lines[0])
	}
}
